### PR TITLE
[no code] basic version for ez launch collection and mint

### PR DESCRIPTION
**Pull request by jillxuu**

### CODE REVIEW

File: ez-launch/sources/ez_launch.move [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+        while (i < length) {
+            let token = pre_mint_helper(
+                creator,
+                collection,
+                *vector::borrow(&token_description_vec, i),
+                *vector::borrow(&token_name_vec, i),
+                *vector::borrow(&token_uri_vec, i),
+                collection_address,
+            );
+            vector::push_back(&mut collection_details.to_be_minted_vec, object::object_address<Token>(&token));
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:202]: Why not store the `Object<Token>` itself?

File: ez-launch/sources/ez_launch.move [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+
+        assert!(exists<CollectionDetails>(object_signer_address), error::invalid_state(ECOLLECTION_DETAILS_DOES_NOT_EXIST));
+        let collection_details_obj = borrow_global_mut<CollectionDetails>(object_signer_address);
+        vector::for_each_ref(&collection_details_obj.coin_payments, |coin_payment| {
+            let coin_payment: &CoinPayment<AptosCoin> = coin_payment;
+            coin_payment::execute(user, coin_payment);
+        });
+
+        let index = if (collection_details_obj.random_mint) {
+            timestamp::now_seconds() % vector::length<address>(&collection_details_obj.to_be_minted_vec)
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:228]: We don't need the `<address>` type generics here, it's all inferred.

File: token-minter/sources/modules/token_components.move [hunk 3]
```diff
@@ -201,6 +203,12 @@ module minter::token_components {
         property_map::remove(&authorized_borrow(creator, token).property_mutator_ref, &key);
     }
 
+    public fun disable_ungated_transfer<T: key>(owner: &signer, token: Object<T>) acquires TokenRefs {
```
Comment (JohnChangUK) [inline token-minter/sources/modules/token_components.move:206]: We can remove this. Unless we want to be able to disable and enable ungated transfer.

File: ez-launch/sources/ez_launch.move [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+    public entry fun create_collection(
+        creator: &signer,
+        description: String,
+        name: String,
+        uri: String,
+        mutable_collection_metadata: bool, // including description, uri, royalty, to make creator life easier lol
+        mutable_token_metadata: bool, // including description, name, properties, uri, to make creator life easier lol
+        random_mint: bool,
+        is_soulbound: bool,
+        token_name_vec: vector<String>, // not provided by creator, we could parse from metadata json file, this vec need to be ordered if order is enforced
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:68]: good call, im gonna separate out the `pre_mint` into a separate func so that it can be batch called

File: ez-launch/sources/ez_launch.move [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    use aptos_token_objects::collection::Collection;
+    use aptos_token_objects::royalty;
+    use aptos_token_objects::token;
+    use aptos_token_objects::token::Token;
+    use minter::coin_payment::{Self, CoinPayment};
+    use minter::collection_components;
+    use minter::token_components;
+    use minter::transfer_token;
+
+    const ENOT_OWNER: u64 = 1;
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:38]: ah curious how does this comment get surfaced in petra? like a code snippet or sth?

[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:38]: interesting til!

File: ez-launch/sources/ez_launch.move [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+        if (collection_details_obj.is_soulbound) {
+            assert!(token_components::contains_token_refs(token_address), ETOKEN_REFS_DOES_NOT_EXIST);
+            token_components::disable_ungated_transfer<Token>(&object_signer, token);
+        };
+
+        token
+    }
+
+    public fun pre_mint_helper(
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:245]: i can put all the test cases to the same file so that i don't need to worry about not getting the object from some helper calls?

File: ez-launch/sources/ez_launch.move [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+            is_soulbound,
+        });
+
+        let object_addr = signer::address_of(&object_signer);
+        let collection_details = borrow_global_mut<CollectionDetails>(object_addr);
+        
+        if (option::is_some(&mint_fee)) {
+            let mint_fee_category = b"Mint Fee";
+            let coin_payment = coin_payment::create<AptosCoin>(option::extract(&mut mint_fee), creator_addr, string::utf8(mint_fee_category));
+            vector::push_back(&mut collection_details.coin_payments, coin_payment);
```
[the remark this answers is not in the document]
Comment (JohnChangUK) [reply ez-launch/sources/ez_launch.move:182]: Each `CoinPayment<T>` is type bound by the coin used. If the user wants to support multiple payments, they would need to store reference to different types of `CoinPayment<T>`, i.e. in a struct. So either way I think it's the same. LMK

File: ez-launch/sources/ez_launch.move [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+        let index = if (collection_details_obj.random_mint) {
+            timestamp::now_seconds() % vector::length<address>(&collection_details_obj.to_be_minted_vec)
+        } else {
+            0
+        };
+        let token_address = *vector::borrow<address>(&collection_details_obj.to_be_minted_vec, index);
+        let token = object::address_to_object<Token>(token_address);
+        let user_address = signer::address_of(user);
+        object::transfer(&object_signer, token, user_address);
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:235]: hmm might be tricky in this use case cuz we pre-mint to the collection address, and `transfer_token::transfer`  need to take in token constructor ref which we don't have when user minting

[the remark this answers is not in the document]
Comment (JohnChangUK) [reply ez-launch/sources/ez_launch.move:235]: Makes sense, stick with the transfer from the object module :)

File: ez-launch/sources/ez_launch.move (hunk 4 above)
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:68]: i wonder would compressed nfts solve this for all 🤔🤔

File: ez-launch/sources/ez_launch.move [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+        let length = vector::length(&token_name_vec);
+        let i = 0;
+        let collection = object::object_from_constructor_ref(&constructor_ref);
+        let collection_address = object::object_address<Collection>(&collection);
+
+        move_to(&object_signer, EZLaunchRefs { extend_ref: object::generate_extend_ref(&constructor_ref) });
+
+        // pre mint all the tokens to the collection address
+        while (i < length) {
+            let token = pre_mint_helper(
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:194]: agree on pre-mint as we separate entry call to call against, not coupled with collection creation, but dont think it should be designed this way as it leaves out a key piece, where are token metadata passed in? if its passed in by this func, only creators can call, otherwise anyone could dump trash tokens to collections. 

other than that, the proposal makes sense with keeping track of the stages and separate pre_mint call. 

btw it's also challenging to track whether all tokens are pre-minted successfully, as collection could be unlimited, so we should leave to sdk to batch call this func iteratively instead

File: ez-launch/sources/ez_launch.move [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        let collection_details_obj = borrow_global_mut<CollectionDetails>(object_signer_address);
+        vector::for_each_ref(&collection_details_obj.coin_payments, |coin_payment| {
+            let coin_payment: &CoinPayment<AptosCoin> = coin_payment;
+            coin_payment::execute(user, coin_payment);
+        });
+
+        let index = if (collection_details_obj.random_mint) {
+            timestamp::now_seconds() % vector::length<address>(&collection_details_obj.to_be_minted_vec)
+        } else {
+            0
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:230]: i guess the order of the token if named numerically would be hard to enforce for minting anyway, this performance gain is reasonable.

File: token-minter/sources/modules/token_components.move (hunk 3 above)
Comment (jillxuu) [reply token-minter/sources/modules/token_components.move:206]: it is unfortunately needed as when we pre-mint as creator, need to enable ungated transfer, and then disable it later if token is soulbound after user claims/mints.

File: ez-launch/sources/ez_launch.move [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+            let coin_payment = coin_payment::create<AptosCoin>(option::extract(&mut mint_fee), creator_addr, string::utf8(mint_fee_category));
+            vector::push_back(&mut collection_details.coin_payments, coin_payment);
+        };
+
+        let length = vector::length(&token_name_vec);
+        let i = 0;
+        let collection = object::object_from_constructor_ref(&constructor_ref);
+        let collection_address = object::object_address<Collection>(&collection);
+
+        move_to(&object_signer, EZLaunchRefs { extend_ref: object::generate_extend_ref(&constructor_ref) });
```
[the remark this answers is not in the document]
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:190]: `collection_object_signer` would require a creator signer to authorize the borrow, but in user minting, we do not need to pass in creator signer at all. so we need a unauthorized borrow to get the refs. so need to create it in the new module

File: ez-launch/sources/ez_launch.move [hunk 12]
```diff
@@ -0,0 +1,413 @@
...
+        tokens_burnable_by_creator: bool,
+        tokens_transferable_by_creator: bool,
+        max_supply: Option<u64>,
+        mint_fee: Option<u64>,
+        royalty_numerator: Option<u64>,
+        royalty_denominator: Option<u64>,
+    ): Object<Collection> acquires CollectionDetails {
+        let creator_addr = signer::address_of(creator);
+
+        assert!(option::is_some(&royalty_numerator) == option::is_some(&royalty_denominator), error::invalid_argument(EROYALTY_CONFIGURATION_INVALID));
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:189]: Why is this asserting if the numerator and denominator is the same?

File: ez-launch/sources/ez_launch.move [hunk 13]
```diff
@@ -0,0 +1,413 @@
...
+        let royalty = if (option::is_some(&royalty_numerator) && option::is_some(&royalty_denominator)) {
+            option::some(royalty::create(option::extract(&mut royalty_numerator), option::extract(&mut royalty_denominator), creator_addr))
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:191]: We need checks on if the royalty passed in is zero or not. This should not be allowed. Ideally, I rather pass in u64 rather than `Option`. Kevin also mentioned that there were problems passing in `Option` in the past.

File: ez-launch/sources/ez_launch.move [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+            mutable_collection_metadata, // mutable_uri
+            mutable_token_metadata, // mutable_token_description
+            mutable_token_metadata, // mutable_token_name
+            mutable_token_metadata, // mutable_token_properties
+            mutable_token_metadata, // mutable_token_uri
+            mutable_collection_metadata, // mutable_royalty
+            tokens_burnable_by_creator,
+            tokens_transferable_by_creator,
+        );
+        collection_components::init_collection_properties(&constructor_ref, collection_properties);
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:232]: We can replace all this with 1 line in this PR: https://github.com/aptos-labs/token-minter/pull/32

:)

File: ez-launch/sources/ez_launch.move [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+    }
+
+    // ================================= Helper  ================================= //
+
+    fun create_collection_internal(
+        creator: &signer,
+        description: String,
+        name: String,
+        uri: String,
+        mutable_collection_metadata: bool,
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:176]: We won't need these also as part of https://github.com/aptos-labs/token-minter/pull/32

File: ez-launch/sources/ez_launch.move [hunk 16]
```diff
@@ -0,0 +1,413 @@
...
+            coin_payments: vector[],
+            random_mint,
+            is_soulbound,
+            ready_to_mint: false,
+        });
+
+        let object_addr = signer::address_of(&object_signer);
+        let collection_details = borrow_global_mut<CollectionDetails>(object_addr);
+        
+        if (option::is_some(&mint_fee)) {
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:247]: Let's have this in it's own function, i.e. `add_mint_fee`.

File: ez-launch/sources/ez_launch.move [hunk 17]
```diff
@@ -0,0 +1,413 @@
...
+    }
+
+    fun token_pre_mint_internal(
+        creator: &signer,
+        collection: Object<Collection>,
+        description: String,
+        name: String,
+        uri: String,
+    ): Object<Token> {
+        assert_owner<Collection>(signer::address_of(creator), collection);
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:269]: Remove all typed generics in this file, they ain't needed :D

File: ez-launch/sources/ez_launch.move [hunk 18]
```diff
@@ -0,0 +1,413 @@
...
+        let constructor_ref = &token::create(
+            creator,
+            collection::name(collection),
+            description,
+            name,
+            royalty::get(collection),
+            uri,
+        );
+
+        let collection_address = object::object_address<Collection>(&collection);
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:279]: `<Collection>` can go from earlier comment

File: ez-launch/sources/ez_launch.move [hunk 19]
```diff
@@ -0,0 +1,413 @@
...
+        object::object_from_constructor_ref(constructor_ref)
+    }
+    
+    fun mint_internal(
+        user: &signer,
+        collection: Object<Collection>,
+    ): Object<Token> acquires CollectionDetails, EZLaunchRefs {
+        let object_signer = collection_signer(collection);
+        let object_signer_address = signer::address_of(&object_signer);
+        assert!(exists<CollectionDetails>(object_signer_address), error::invalid_state(ECOLLECTION_DETAILS_DOES_NOT_EXIST));
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:292]: Let's break line after an `assert`. This makes it a lot easier to read.

File: ez-launch/sources/ez_launch.move [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+        let collection_details_obj = borrow_global_mut<CollectionDetails>(object_signer_address);
+        let to_be_minted = collection_details_obj.to_be_minted;
+        let length = vector::length(&to_be_minted);
+
+        assert!(length > 0, error::permission_denied(EMINTING_HAS_ENDED));
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:297]: I think a better error would be `ETOKENS_ALL_MINTED`.

File: ez-launch/sources/ez_launch.move [hunk 21]
```diff
@@ -0,0 +1,413 @@
...
+        assert!(collection_details_obj.ready_to_mint, error::permission_denied(EMINTING_HAS_NOT_STARTED_YET));
+
+        // (jill) check against whitelist
+
+        vector::for_each_ref(&collection_details_obj.coin_payments, |coin_payment| {
+            let coin_payment: &CoinPayment<AptosCoin> = coin_payment;
+            coin_payment::execute(user, coin_payment);
+        });
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:305]: This can be in a separate function.

File: ez-launch/sources/ez_launch.move [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+
+        
+        let index = if (collection_details_obj.random_mint) {
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:308]: Move this to a separate function, very hard to read at first glance.

File: ez-launch/sources/ez_launch.move [hunk 23]
```diff
@@ -0,0 +1,413 @@
...
+        let token_address = object::object_address(&token);
+        let user_address = signer::address_of(user);
+        object::transfer(&object_signer, token, user_address);
+        vector::pop_back(&mut to_be_minted);
+
+        if (collection_details_obj.is_soulbound) {
+            assert!(token_components::contains_token_refs(token_address), ETOKEN_REFS_DOES_NOT_EXIST);
+            token_components::disable_ungated_transfer<Token>(&object_signer, token);
+        };
+        token
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:323]: Move this to a separate function, very hard to read at first glance.

File: ez-launch/sources/ez_launch.move [hunk 24]
```diff
@@ -0,0 +1,413 @@
...
+        assert!(
+            exists<EZLaunchRefs>(collection_address),
+            error::not_found(EEZ_LAUNCH_REFS_DOES_NOT_EXIST)
+        );
+        borrow_global<EZLaunchRefs>(collection_address)
+    }
+
+    // ================================= Tests  ================================= //
+
+    #[test_only]
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:356]: Let's have tests in a separate directory under `tests`. We shouldn't be writing tests in the same file now, as we now can write tests in a `tests` directory. 

This also makes it easier to read and reason about the module.

File: ez-launch/sources/ez_launch.move [hunk 25]
```diff
@@ -0,0 +1,413 @@
...
+    use std::string::utf8;
+    #[test_only]
+    use ez_launch::ez_launch;
+
+    #[test(creator = @0x1)]
+    fun test_create_collection(creator: &signer) acquires CollectionDetails {
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:362]: These tests are quite basic, there are no edge cases tested here. The tests will grow over time, hence we should have this in a separate file and directory. 

Make a TODO to add more test cases in.

File: ez-launch/sources/ez_launch.move [hunk 26]
```diff
@@ -0,0 +1,413 @@
...
+    const ETOKEN_METADATA_CONFIGURATION_INVALID: u64 = 7;
+    /// Token Minting has not yet started.
+    const EMINTING_HAS_NOT_STARTED_YET: u64 = 8;
+    /// Token Minting has ended.
+    const EMINTING_HAS_ENDED: u64 = 9;
+
+    #[resource_group_member(group = aptos_framework::object::ObjectGroup)]
+    struct CollectionDetails has key {
+        coin_payments: vector<CoinPayment<AptosCoin>>,
+        to_be_minted: vector<Object<Token>>, // to be minted token
```
Comment (JohnChangUK) [inline ez-launch/sources/ez_launch.move:62]: Would `available_tokens` be a better name for this?

File: ez-launch/sources/ez_launch.move (hunk 24 above)
Comment (jillxuu) [reply ez-launch/sources/ez_launch.move:356]: we're having a lot of entry functions here which don't return objects at all, so we have couple internal funcs that are does return object for us to test against. thats why they are in the same file as i don't want to expose internal funcs. any better suggestions on that?

File: ez-launch/sources/ez_launch.move [hunk 27]
```diff
@@ -0,0 +1,462 @@
...
+    ): Option<Royalty> {
+        assert!(option::is_some(royalty_numerator) == option::is_some(royalty_denominator), error::invalid_argument(EROYALTY_CONFIGURATION_INVALID));
+        if (option::is_some(royalty_numerator) && option::is_some(royalty_denominator) && option::extract(royalty_numerator) != 0 && option::extract(royalty_denominator) != 0) {
+            option::some(royalty::create(option::extract(royalty_numerator), option::extract(royalty_denominator), creator_addr))
+        } else {
+            option::none()
+        }
+    }
+
+    fun configure_collection_and_token_properties(
```
Comment (jillxuu) [inline ez-launch/sources/ez_launch.move:347]: man this is still so ugly lol. @JohnChangUK

Comment (JohnChangUK) [reply ez-launch/sources/ez_launch.move:347]: Ah I see. But the whole point of default values was so that the user does not have to provide these at the time of creating a collection. I think it should be in separate function calls, to split out the creation of the collection vs the properties.

Better yet, if the user provides these details, we should not be checking if it is true or false, but just setting them all directly, by calling each setter. This means that the user has already decided to set the collection properties and once configured, should be immutable :o

File: ez-launch/sources/ez_launch.move (hunk 24 above)
Comment (JohnChangUK) [reply ez-launch/sources/ez_launch.move:402]: Internal functions that return objects can simply be public functions, and you call the public functions from the test file :)

**Review by JohnChangUK — APPROVED** [review]
